client/src/main/java/io/pravega/client/byteStream/ByteStreamWriter.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.byteStream;
+
+import io.pravega.client.state.RevisionedStreamClient;
+import io.pravega.client.stream.EventStreamWriter;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+
+/**
+ * Allows for writing raw bytes directly to a segment. This is intended as low level building block
+ * for creating higher level components and is not appropriate for most applications to use
+ * directly.
+ * 
+ * This class does not frame, attach headers, or otherwise modify the bytes written to it in any
+ * way. So unlike {@link EventStreamWriter} or {@link RevisionedStreamClient} the data written
+ * cannot be split apart when read. As such, any bytes written by this API can ONLY be read using
+ * {@link ByteStreamReader}. Similarly, unless some sort of framing is added it is probably an error
+ * to have multiple ByteStreamWriters write to the same segment as this will result in interleaved
+ * data.
+ * 
+ * The methods on this class are non-blocking unless otherwise specified. As such data passed to a
+ * {@link #write(byte[])} call cannot be assumed to be persisted until a flush has been called.
+ * 
+ * It is safe to invoke methods on this class from multiple threads but doing so will not result in
+ * an increase in performance.
+ */
+public abstract class ByteStreamWriter extends OutputStream {
```

Comment:
In this current PR, we are writing to a single segment. Say that in the future I want to benefit of parallelism and I want to upload different chunks to different segments in parallel. Would that be an extension of this class?

Reply:
We don't generally extend interfaces. It's better to just create a new interface that might share code under the hood.

Reply:
I have created this issue to track the discussion: #3010.